bindings/haskell/src/logger.rs

```diff
@@ -0,0 +1,36 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::ffi::{c_char, CString};
+
+pub struct HsLogger {
+    pub callback: extern "C" fn(*const c_char),
+}
+
+impl log::Log for HsLogger {
```

Comment:
I don't like the idea of exposing a `HsLogger`in hashell binding directly. My current understanding is:

- Provide a `HaskellLayer` which implements `opendal::Layer` in rust.
- `HaskellLayer` should expose FFI to haskell like a trait in haskell so other libs can implement it (I don't is it a thing in haskell).
- And finnally, users can implement whatever they want in haskell.

So: haskell binding itself should know nothing about logging

Reply:
> I don't like the idea of exposing a HsLoggerin hashell binding directly.

If you don't like exposing a `HsLogger`, we can also implement `logging` with pure haskell without any rust code. My purpose of using `HsLogger` is to reduce duplicate code and maintain consistency between rust logs and Haskell logs.


> My current understanding is:
> - Provide a HaskellLayer which implements opendal::Layer in rust.
> - HaskellLayer should expose FFI to haskell like a trait in haskell so other libs can implement it (I don't is it a thing in haskell).
> - And finnally, users can implement whatever they want in haskell.

My understanding is that `OpMonad` in haskell side provides almost same function with `HaskellLayer`. As result, it's not necessary to provide a `HaskellLayer`. Users can implement whatever they want through `OpMonad`.


> So: haskell binding itself should know nothing about logging

Users might need to integrate OpenDAL logging with their own logging system. I don't think it's a good idea.

Reply:
> My understanding is that `OpMonad` in haskell side provides almost same function with `HaskellLayer`. As result, it's not necessary to provide a `HaskellLayer`. Users can implement whatever they want through `OpMonad`.

Ok, seems we don't need to extra work. Maybe we just need an example for haskell users that how can they integrate their logging system with opendal.

Reply:
> I don't get it. It's exactly what I want: users can use their own logging system.

When researching haskell logging system, I found that hakell logging ecosystem is more diverse, and there isn't a universally-adopted standard like the log crate in rust. As result, I design simplest callback function logging system which does not rely on any existing logging framework. We can add some examples for popular logging system like `hslogger`, `katip`.

Reply:
> As result, I design simplest callback function logging system which does not rely on any existing logging framework.

I don't think we need this. Users should pick and setup logging system on their needs.